src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets

```diff
@@ -161,19 +161,45 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <_FrameworkIdentifierForImplicitDefine Condition=" '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and '$(_FrameworkVersionForImplicitDefine)' >= '5.0' ">NET</_FrameworkIdentifierForImplicitDefine>
+    <VersionlessImplicitFrameworkDefine Condition=" '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and '$(_FrameworkVersionForImplicitDefine)' >= '5.0' ">$(VersionlessImplicitFrameworkDefine);$(_FrameworkIdentifierForImplicitDefine)</VersionlessImplicitFrameworkDefine>
     <_FrameworkIdentifierForImplicitDefine Condition=" '$(TargetFrameworkIdentifier)' == '.NETFramework'">NET</_FrameworkIdentifierForImplicitDefine>
 
-    <_FrameworkVersionForImplicitDefine>$(TargetFrameworkVersion.TrimStart('vV'))</_FrameworkVersionForImplicitDefine>
-
     <_FrameworkVersionForImplicitDefine>$(_FrameworkVersionForImplicitDefine.Replace('.', '_'))</_FrameworkVersionForImplicitDefine>
-    
+
     <_FrameworkVersionForImplicitDefine Condition=" '$(TargetFrameworkIdentifier)' == '.NETFramework'">$(_FrameworkVersionForImplicitDefine.Replace('_', ''))</_FrameworkVersionForImplicitDefine>
 
     <ImplicitFrameworkDefine>$(_FrameworkIdentifierForImplicitDefine)$(_FrameworkVersionForImplicitDefine)</ImplicitFrameworkDefine>
   </PropertyGroup>
 
+  <!-- Add conditional compilation symbols for the target platform (for example ANDROID, IOS, WINDOWS) -->
+  <PropertyGroup Condition=" '$(DisableImplicitFrameworkDefines)' != 'true' and '$(TargetPlatformIdentifier)' != ''">
+    <_PlatformIdentifierForImplicitDefine>$(TargetPlatformIdentifier.ToUpperInvariant())</_PlatformIdentifierForImplicitDefine>
+    <_PlatformVersionForImplicitDefine>$(TargetPlatformVersion.Replace('.', '_'))</_PlatformVersionForImplicitDefine>
+    <ImplicitPlatformDefine>$(_PlatformIdentifierForImplicitDefine);$(_PlatformIdentifierForImplicitDefine)$(_PlatformVersionForImplicitDefine)</ImplicitPlatformDefine>
+    
+    <ImplicitFrameworkDefine>$(ImplicitFrameworkDefine);$(ImplicitPlatformDefine)</ImplicitFrameworkDefine>
+  </PropertyGroup>
+  
+  <!-- Add conditional compilation symbols for target framework backwards compatibility with .NET 5.0 and later (for example .NET 6.0 defines NET5_0 and NETCOREAPP3_1) -->
+  <Target Name="GenerateNETCompatibleDefineConstants"
+          Condition=" '$(DisableImplicitFrameworkDefines)' != 'true' and '$(TargetFrameworkIdentifier)' != '' and '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and 
+                      '$(_FrameworkVersionNumberForImplicitDefine)' != '' and '$(_FrameworkVersionNumberForImplicitDefine)' >= '5.0' "
+          BeforeTargets="Build" >
+    <ItemGroup>
+      <_NetCoreAppFrameworkVersions Include="@(SupportedNETCoreAppTargetFramework->'%(Identity)'->TrimStart('.NETCoreApp,Version=v'))" />
+      <_NetCoreAppCompatibleVersions Include="@(_NetCoreAppFrameworkVersions)" Condition=" '%(Identity)' > '3.1' and '$(_FrameworkVersionNumberForImplicitDefine)' > '%(Identity)' " />
```

Comment:
Use version comparison intrinsics for these comparisons.  Also I think it should be comparing against TargetFrameworkVersion instead of _FrameworkVersionNumberForImplicitDefine.